pkgs/desktops/pantheon/apps/elementary-terminal/partly-revert-ad127db.patch

```diff
@@ -0,0 +1,22 @@
+diff --git a/src/Widgets/TerminalWidget.vala b/src/Widgets/TerminalWidget.vala
+index 63341221..a880eeb1 100644
+--- a/src/Widgets/TerminalWidget.vala
++++ b/src/Widgets/TerminalWidget.vala
+@@ -307,12 +307,17 @@ namespace Terminal {
+         public void term_ps () {
+             killed = true;
+ 
++            int tries = 0;
+             /* Check if the shell process is still alive by sending 0 signals */
+             while (Posix.kill (this.child_pid, 0) == 0) {
+                 Posix.kill (this.child_pid, Posix.Signal.HUP);
+                 Posix.kill (this.child_pid, Posix.Signal.TERM);
+                 Thread.usleep (100);
++                if (++tries >= 5)
++                    break;
+             }
++            if (Posix.kill (this.child_pid, 0) != -1)
++                Posix.kill (this.child_pid, Posix.Signal.KILL);
```

Comment:
Not sure how this would help against zombies – killing those cannot really do anything as they are already dead, only not reaped by their parent process.

I also would not this break stuff like [`nohup`](https://en.wikipedia.org/wiki/Nohup)? Though it was probably already the case.

Reply:
Yeah I think we can remove these two lines. Does nohup still affected in 26217a89ed9478c12667ff16ce879da48120a9fd? I actually thought term_ps only handle the shell itself and I though nohup will transfer the protected process when it sees the parent go away.

Reply:
I am not really sure, my POSIX knowledge is pretty superficial. Though I would expect that if a process would be harmed by `SIGKILL`, it would already affected by `SIGTERM`, unless `nohup` masks one but not the other. But reading the info page, it looks like it only masks `SIGHUP`.

Reply:
Testing `nohup yes > /dev/null &` and closing the tab, the process keeps running even with the `SIGKILL` so it probably is not considered a child process. When I run it without `&` I get a confirmation prompt upon trying to close the tab and clicking close will indeed terminate it. So I guess the `SIGKILL` will be only used for this prompt and it makes sense there.

Reply:
Hmm, I am not sure what does send `SIGHUP` to other jobs on terminal closing, though. Maybe the shell receives `SIGHUP` and `SIGTERM` from this and then invokes `SIGHUP` on all its jobs? Or it is handled by VTE? Weirdly, if I run just `ping google.com &`, closing the tab does not terminate the process, even though it does not mask `SIGHUP` (I can kill it with `pkill ping -HUP`). :confused:

Reply:
> When I run it without `&` ...

Yeah from my testing it will be considered as a foreground process and will be handled in [`kill_fg`](https://github.com/elementary/terminal/blob/5a9ac9197e89191c1477a3ae0d7d694dd152365b/src/Widgets/TerminalWidget.vala#L304).

> if I run just `ping google.com &`, closing the tab does not terminate the proces

I think this is possible if `gsettings get io.elementary.terminal.settings save-exited-tabs` is `true` (which is the default) and multiple tabs are opened.

Reply:
I only had a single tab opened and after closing it elementary terminal process exited and the `ping` process moved under `systemd --user`

Reply:
> the ping process moved under systemd --user

Weird, on a 26217a89ed9478c12667ff16ce879da48120a9fd GNOME VM I can only reproduce this on gnome-console but not on elementary-terminal :confused: